### PR TITLE
CSI-2778 reducing the server file length

### PR DESCRIPTION
**Pull request by zingero**

### CODE REVIEW

File: controller/controller_server/exception_handler.py [hunk 1]
```diff
@@ -4,17 +4,29 @@
...
+    array_errors.SnapshotSourcePoolMismatch: grpc.StatusCode.INVALID_ARGUMENT,
     array_errors.ObjectNotFoundError: grpc.StatusCode.NOT_FOUND,
+    array_errors.HostNotFoundError: grpc.StatusCode.NOT_FOUND,
+    array_errors.NoIscsiTargetsFoundError: grpc.StatusCode.NOT_FOUND,
+    array_errors.VolumeAlreadyUnmappedError: grpc.StatusCode.NOT_FOUND,
+    ObjectIdError: grpc.StatusCode.NOT_FOUND,
     array_errors.PermissionDeniedError: grpc.StatusCode.PERMISSION_DENIED,
-    array_errors.NotEnoughSpaceInPool: grpc.StatusCode.RESOURCE_EXHAUSTED
+    array_errors.ObjectIsStillInUseError: grpc.StatusCode.FAILED_PRECONDITION,
+    array_errors.VolumeMappedToMultipleHostsError: grpc.StatusCode.FAILED_PRECONDITION,
```
Comment (oriyarde) [inline controller/controller_server/exception_handler.py:25]: this error seems to be relevant only to a specific method.
if so, then as I [wrote before](https://github.com/IBM/ibm-block-csi-driver/pull/305#discussion_r608638835), please ensure there are no "method-specific" errors here

File: controller/controller_server/csi_controller_server.py [hunk 2]
```diff
@@ -335,33 +260,14 @@ def ControllerUnpublishVolume(self, request, context):
 
             logger.info("finished ControllerUnpublishVolume")
             return csi_pb2.ControllerUnpublishVolumeResponse()
-
```
[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/csi_controller_server.py:338]: Is it a part of our convention?

File: controller/controller_server/exception_handler.py (hunk 1 above)
Comment (oriyarde) [reply controller/controller_server/exception_handler.py:25]: @ArbelNathan what do you think?

File: controller/controller_server/csi_controller_server.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:338]: seems to me, yes

File: controller/controller_server/csi_controller_server.py [hunk 3]
```diff
@@ -123,9 +117,9 @@ def CreateVolume(self, request, context):
                 logger.info("finished create volume")
                 return res
         except array_errors.InvalidArgumentError as ex:
-            handle_exception(ex, context, grpc.StatusCode.INVALID_ARGUMENT, csi_pb2.CreateVolumeResponse)
+            return handle_exception(ex, context, grpc.StatusCode.INVALID_ARGUMENT, csi_pb2.CreateVolumeResponse)
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/controller_server/csi_controller_server.py:120]: it's not progressing and won't soo.

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:120]: then @zingero please add test coverage here, at least for these fixes

File: controller/controller_server/exception_handler.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/controller_server/exception_handler.py:25]: well, what do you suggest?

File: controller/controller_server/csi_controller_server.py [hunk 4]
```diff
@@ -174,13 +168,12 @@ def _handle_existing_volume_source(self, volume, source_id, source_type, context
...
         if volume_copy_source_id == source_id:
-            return self._handle_volume_exists_with_same_source(context, source_id, source_type, volume_name, volume)
+            return self._handle_volume_exists_with_same_source(source_id, source_type, volume_name, volume)
         return self._handle_volume_exists_with_different_source(context, source_id, source_type, volume_name)
 
-    def _handle_volume_exists_with_same_source(self, context, source_id, source_type, volume_name, volume):
+    def _handle_volume_exists_with_same_source(self, source_id, source_type, volume_name, volume):
         logger.debug(
             "Volume {0} exists and it is a copy of {1} {2}.".format(volume_name, source_type, source_id))
-        context.set_code(grpc.StatusCode.OK)
```
[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/csi_controller_server.py:183]: OK is the default status code. No other status code is set before reaching this line.

File: controller/controller_server/exception_handler.py (hunk 1 above)
Comment (oriyarde) [reply controller/controller_server/exception_handler.py:25]: keep the try-except wherever there is uncommon exceptions to except, as `CreateVolume` is in current develop branch

File: controller/controller_server/csi_controller_server.py [hunk 5]
```diff
@@ -190,136 +183,68 @@ def _handle_volume_exists_with_different_source(self, context, source_id, source
...
         logger.info("ControllerUnpublishVolume")
         try:
-            try:
-                utils.validate_unpublish_volume_request(request)
-            except ValidationException as ex:
-                logger.exception(ex)
-                context.set_details(ex.message)
-                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
-                return csi_pb2.ControllerUnpublishVolumeResponse()
+            utils.validate_unpublish_volume_request(request)
```
[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/csi_controller_server.py]: Done

File: controller/controller_server/csi_controller_server.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/csi_controller_server.py]: Done

File: controller/controller_server/exception_handler.py [hunk 6]
```diff
@@ -4,17 +4,29 @@
...
+    array_errors.LunAlreadyInUseError: grpc.StatusCode.RESOURCE_EXHAUSTED,
+    array_errors.NoAvailableLunError: grpc.StatusCode.RESOURCE_EXHAUSTED,
+    array_errors.NotEnoughSpaceInPool: grpc.StatusCode.RESOURCE_EXHAUSTED,
+    array_errors.SnapshotAlreadyExists: grpc.StatusCode.ALREADY_EXISTS,
 }
```
[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/exception_handler.py:37]: Done

File: controller/controller_server/exception_handler.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/controller_server/exception_handler.py:29]: ok, sounds good

File: controller/controller_server/csi_controller_server.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/csi_controller_server.py:120]: Done

File: controller/controller_server/exception_handler.py (hunk 1 above)
Comment (zingero) [reply controller/controller_server/exception_handler.py:29]: @ArbelNathan, Ori's suggestion adds more lines and indentations to the already huge file.
This is the reason for removing it.
Do you still prefer his solution? If so, please explain.

File: controller/controller_server/csi_controller_server.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:183]: ok. how did you verify that this is the default btw?

[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/csi_controller_server.py:183]: By running tests

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:183]: unit? community? automation?

[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/csi_controller_server.py:183]: I ran our unit tests and automation tests (both passed)
By running the unit tests I verified the default status code.

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/controller_server/csi_controller_server.py:183]: our unit tests use a [fake context object](https://github.com/IBM/ibm-block-csi-driver/blob/eb9b23e4a25656667e4767cf997d7471cf451d0c/controller/tests/utils.py#L34), so I wouldn't rely on them for this.
automation doesn't assert our status codes at all

[the remark this answers is not in the document]
Comment (zingero) [reply controller/controller_server/csi_controller_server.py]: I reverted this line. I'll remove it later, once we have more time

File: controller/controller_server/exception_handler.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/controller_server/exception_handler.py:29]: the `grpc StatusCode` we are returning are by spec, if a grpc is not supposed to return by spec, we should follow.

Comment (zingero) [reply controller/controller_server/exception_handler.py:29]: After talking with Arbel, I decided to change the code. Currently, we have urgent things to do. Thus, I postpone this task

**Pull request closed without merging**